### PR TITLE
Findmax enhancements

### PR DESCRIPTION
**Pull request by MarkWolters**
Adding Findmax as a separate command rather than an option of the Optimize command, although the original still exists also. This version of findmax can be used to optimize on either rate or on threads and more parameters can be added with minimal code changes.

### CODE REVIEW

File: nbr/src/main/java/io/nosqlbench/scenarios/simframe/optimizers/findmax/CMD_findmax.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2023 nosqlbench
```
Comment (msmygit) [inline nbr/src/main/java/io/nosqlbench/scenarios/simframe/optimizers/findmax/CMD_findmax.java:2]: #nit 
```suggestion
 * Copyright (c) 2020-2024 nosqlbench
```

File: nbr/src/main/java/io/nosqlbench/scenarios/simframe/optimizers/findmax/FindmaxAnalyzer.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2024 nosqlbench
```
Comment (msmygit) [inline nbr/src/main/java/io/nosqlbench/scenarios/simframe/optimizers/findmax/FindmaxAnalyzer.java:2]: #nit 
```suggestion
 * Copyright (c) 2020-2024 nosqlbench
```

File: nbr/src/main/java/io/nosqlbench/scenarios/simframe/optimizers/findmax/FindmaxConfig.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (c) 2023 nosqlbench
```
Comment (msmygit) [inline nbr/src/main/java/io/nosqlbench/scenarios/simframe/optimizers/findmax/FindmaxConfig.java:2]: #nit 
```suggestion
 * Copyright (c) 2020-2024 nosqlbench
```

File: nbr/src/main/java/io/nosqlbench/scenarios/simframe/optimizers/findmax/FindmaxFrameFunction.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2023 nosqlbench
```
Comment (msmygit) [inline nbr/src/main/java/io/nosqlbench/scenarios/simframe/optimizers/findmax/FindmaxFrameFunction.java:2]: #nit 
```suggestion
 * Copyright (c) 2020-2024 nosqlbench
```

File: nbr/src/main/java/io/nosqlbench/scenarios/simframe/optimizers/findmax/FindmaxFrameFunction.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        this.settings = settings;
+        this.flywheel = flywheel;
+        this.capture = capture;
+        this.journal = journal;
+        this.model = model;
+    }
+
+    @Override
+    public double value(double[] point) {
+        System.out.println("━".repeat(40));
```
Comment (msmygit) [inline nbr/src/main/java/io/nosqlbench/scenarios/simframe/optimizers/findmax/FindmaxFrameFunction.java:53]: #nit

Should these SOPs be converted to actual loggers?

File: nbr/src/main/java/io/nosqlbench/scenarios/simframe/optimizers/findmax/FindmaxFrameParams.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2023 nosqlbench
```
Comment (msmygit) [inline nbr/src/main/java/io/nosqlbench/scenarios/simframe/optimizers/findmax/FindmaxFrameParams.java:2]: #nit 
```suggestion
 * Copyright (c) 2020-2024 nosqlbench
```

File: nbr/src/main/java/io/nosqlbench/scenarios/simframe/optimizers/findmax/FindmaxParamModel.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) 2023 nosqlbench
```
Comment (msmygit) [inline nbr/src/main/java/io/nosqlbench/scenarios/simframe/optimizers/findmax/FindmaxParamModel.java:2]: #nit 
```suggestion
 * Copyright (c) 2020-2024 nosqlbench
```

File: nbr/src/main/java/io/nosqlbench/scenarios/simframe/optimizers/findmax/FindmaxParamModel.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.DoubleConsumer;
+
+public class FindmaxParamModel {
+    private final List<GenericParamModel> params = new ArrayList<>();
+
+    public FindmaxParamModel add(String name, double min, double initial, double max, DoubleConsumer effector) {
+        if (min>initial || initial > max) {
+            throw new RuntimeException("parameters must be in min<initial<max order, but " + name + " was min=" + min + ", max=" + max);
```
Comment (msmygit) [inline nbr/src/main/java/io/nosqlbench/scenarios/simframe/optimizers/findmax/FindmaxParamModel.java:31]:
```suggestion
            throw new RuntimeException("parameters must be in min<initial<max order, but " + name + " was min=" + min + ", initial=" + initial + ", max=" + max);
```

File: nbr/src/main/java/io/nosqlbench/scenarios/simframe/optimizers/findmax/NBFindmaxInfo.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright (c) 2023 nosqlbench
```
Comment (msmygit) [inline nbr/src/main/java/io/nosqlbench/scenarios/simframe/optimizers/findmax/NBFindmaxInfo.java:2]: #nit 
```suggestion
 * Copyright (c) 2020-2024 nosqlbench
```

File: nbr/src/main/java/io/nosqlbench/scenarios/simframe/planning/SimFrameAction.java [hunk 10]
```diff
@@ -0,0 +1,23 @@
+/*
+ * Copyright (c) 2024 nosqlbench
```
Comment (msmygit) [inline nbr/src/main/java/io/nosqlbench/scenarios/simframe/planning/SimFrameAction.java:2]: #nit 
```suggestion
 * Copyright (c) 2020-2024 nosqlbench
```

File: nbr/src/main/java/io/nosqlbench/scenarios/simframe/planning/SimFrameFunctionAnalyzer.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2024 nosqlbench
```
Comment (msmygit) [inline nbr/src/main/java/io/nosqlbench/scenarios/simframe/planning/SimFrameFunctionAnalyzer.java:2]: #nit 
```suggestion
 * Copyright (c) 2020-2024 nosqlbench
```

**Review by msmygit — DISMISSED** [review]

File: nbr/src/main/java/io/nosqlbench/scenarios/simframe/optimizers/findmax/FindmaxFrameFunction.java (hunk 5 above)
Comment (MarkWolters) [reply nbr/src/main/java/io/nosqlbench/scenarios/simframe/optimizers/findmax/FindmaxFrameFunction.java:53]: That's a good point. They are currently mimicking the behavior of the old findmax but I agree they should probably be converted to logging, along with the other stdout statements in the code.

**Review by jshook — APPROVED** [review]

**Pull request merged**